Detectors/TPC/simulation/src/SpaceCharge.cxx

```diff
@@ -0,0 +1,327 @@
+  setOmegaTauT1T2(omegaTau, t1, t2);
+  if (mUseInitialSCDensity)
+    calculateLookupTables();
+}
+
+void SpaceCharge::calculateLookupTables()
+{
+  // Potential, E field and electron distortion and correction lookup tables
+  mLookUpTableCalculator.ForceInitSpaceCharge3DPoissonIntegralDz(mNRBins, mNZSlices, mNPhiBins, 300, 1e-8);
+
+  // Lookup tables for local ion drift along E field
+  if (mSCDistortionType == SCDistortionType::SCDistortionsRealistic) {
+    for (int iphi = 0; iphi < mNPhiBins; ++iphi) {
+      float phi = mCoordPhi[iphi];
+      TMatrixD& matrixDzA = *mMatrixLocalIonDriftDzA[iphi];
+      TMatrixD& matrixDzC = *mMatrixLocalIonDriftDzC[iphi];
+      TMatrixD& matrixDrphiA = *mMatrixLocalIonDriftDrphiA[iphi];
+      TMatrixD& matrixDrphiC = *mMatrixLocalIonDriftDrphiC[iphi];
+      TMatrixD& matrixDrA = *mMatrixLocalIonDriftDrA[iphi];
+      TMatrixD& matrixDrC = *mMatrixLocalIonDriftDrC[iphi];
+      int roc = o2::utils::Angle2Sector(phi);
+      for (int ir = 0; ir < mNRBins; ++ir) {
+        float radius = mCoordR[ir];
+        for (int iz = 0; iz < mNZSlices; ++iz) {
+          // A side
+          float z = mCoordZ[iz];
+          float x0[3] = { radius, phi, z };
+          float x1[3] = { radius, phi, z - mLengthZSlice };
+          double eVector0[3] = { 0.f, 0.f, 0.f };
+          double eVector1[3] = { 0.f, 0.f, 0.f };
+          mLookUpTableCalculator.GetElectricFieldCyl(x0, roc, eVector0);
+          mLookUpTableCalculator.GetElectricFieldCyl(x1, roc, eVector1);
+          matrixDzA(ir, iz) = DvDEoverv0 * (mLengthZSlice / 2.f) * (eVector0[2] + eVector1[2]);
+          matrixDrphiA(ir, iz) = (mLengthZSlice / 2.f) * (eVector0[1] + eVector1[1]) / sEzField;
+          matrixDrA(ir, iz) = (mLengthZSlice / 2.f) * (eVector0[0] + eVector1[0]) / sEzField;
+          // C side
+          x0[2] *= -1;
+          x1[2] *= -1;
+          mLookUpTableCalculator.GetElectricFieldCyl(x0, roc + 18, eVector0);
+          mLookUpTableCalculator.GetElectricFieldCyl(x1, roc + 18, eVector1);
+          matrixDzC(ir, iz) = DvDEoverv0 * (mLengthZSlice / 2.f) * (eVector0[2] + eVector1[2]);
+          matrixDrphiC(ir, iz) = (mLengthZSlice / 2.f) * (eVector0[1] + eVector1[1]) / sEzField;
+          matrixDrC(ir, iz) = (mLengthZSlice / 2.f) * (eVector0[0] + eVector1[0]) / sEzField;
+        }
+      }
+    }
+    mLookUpLocalIonDriftA->CopyFromMatricesToInterpolator();
+    mLookUpLocalIonDriftC->CopyFromMatricesToInterpolator();
+
+    /// TODO: Propagate current SC density along E field by one time bin for next update
+  }
+
+  mInitLookUpTables = true;
+}
+
+void SpaceCharge::updateLookupTables(float eventTime)
+{
+  if (mTimeInit < 0.)
+    mTimeInit = eventTime; // set the time of first initialization
+  if (TMath::Abs(eventTime - mTimeInit) < mLengthTimeSlice)
+    return; // update only after one time bin has passed
+
+  TMatrixD* spaceChargeA[mNPhiBins];
+  TMatrixD* spaceChargeC[mNPhiBins];
```

Comment:
also here use `unique_ptr`.
As we we discussed, `mLookUpTableCalculator` only reads values, but never takes ownership.

Reply:
done